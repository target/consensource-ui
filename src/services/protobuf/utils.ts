import {
  CertificateRegistryPayload,
  ICertificateRegistryPayload,
} from './compiled';

export interface PayloadInfo {
  payloadBytes: string | Buffer | NodeJS.TypedArray | DataView;
  inputs: string[];
  outputs: string[];
}

export const ACTIONS = CertificateRegistryPayload.Action;

/**
 * Encodes a CertificateRegistryPayload message
 * @param message CertificateRegistryPayload message or plain object to encode
 */
export function encodePayload(
  payload: ICertificateRegistryPayload,
): Uint8Array {
  return CertificateRegistryPayload.encode(payload).finish();
}
