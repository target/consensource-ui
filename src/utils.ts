// Util collection

/**
 * Selects keys from an object
 */
export const pluck = (obj: any, ...keys: Array<any>): any =>
    keys.reduce((o, k) => {
        if (typeof obj[k] !== 'undefined') {
            o[k] = obj[k];
        }
        return o;
    }, {});

/**
 * Merges objects, starting from left to right
 */
