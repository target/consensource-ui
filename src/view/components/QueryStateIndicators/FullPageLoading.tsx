import React from 'react';
import { Backdrop, makeStyles, createStyles } from '@material-ui/core';
import { QueryResult } from 'react-query';
import { SpinnerWithLabel } from './SpinnerWithLabel';
import {
  LoadingWithMinDisplay,
  LoadingWithMinDisplayProps,
} from './LoadingWithMinDisplay';
import { WarningIconError } from './WarningIconError';

export interface FullPageLoadingProps<T extends QueryResult<any>> {
  /**
   * **Default**: _"Failed to load"_
   *
   * Error text that is passed to the `<WarningIconError />`
   */
  errorLabel?: string;
  /**
   * **Default**: _"Loading..."_
   *
   * Loading text that is passed to the `<SpinnerWithLabel />`
   */
  loadingLabel?: string;
  queryRes: LoadingWithMinDisplayProps<T>['queryRes'];
  children: LoadingWithMinDisplayProps<T>['children'];
}

const useStyles = makeStyles(({ zIndex }) =>
  createStyles({
    backdrop: {
      zIndex: zIndex.drawer + 1,
      color: '#fff',
    },
    centered: {
      position: 'absolute',
      left: '50%',
      top: '50%',
      transform: 'translateY(-50%)',
    },
  }),
);

/**
 * Wrapper for `<LoadingWithMinDisplay />` to render a generic full
 * screen loading process.
 *
 * While loading, a spinner with a label is displayed over a backdrop.
 *
 * If an error occurs, a `<WarningIconError />` component is rendered.
 */
export const FullPageLoading = <T extends QueryResult<any>>({
  children,
  loadingLabel = 'Loading...',
  errorLabel = 'Failed to load',
  queryRes,
}: FullPageLoadingProps<T>) => {
  const classes = useStyles();

  return (
    <LoadingWithMinDisplay
      queryRes={queryRes}
      errorIndicator={
        <div className={classes.centered}>
          <WarningIconError size="large">{errorLabel}</WarningIconError>
        </div>
      }
      loadingIndicator={
        <Backdrop open className={classes.backdrop}>
          <SpinnerWithLabel>{loadingLabel}</SpinnerWithLabel>
        </Backdrop>
      }
    >
      {children}
    </LoadingWithMinDisplay>
  );
};
