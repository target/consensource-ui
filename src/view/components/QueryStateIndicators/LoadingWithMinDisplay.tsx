import React, { useEffect, useState, useRef } from 'react';
import { CircularProgress } from '@material-ui/core';
import { QueryResult } from 'react-query';
import { WarningIconError } from './WarningIconError';

export interface LoadingWithMinDisplayProps<T extends QueryResult<any>> {
  /**
   * **Default value**: _750_
   *
   * Minimum number of milliseconds to display the loading animation for.
   * Note that the timer will only begin once the wait timer has expired.
   *
   */
  minDisplayTimeMs?: number;
  /**
   * **Default value**: _250_
   *
   * Number of milliseconds to wait before displaying the loading animation.
   * Used to prevent the loading animation from flashing on screen very briefly
   * if loading completes in less than `waitTimeMs` milliseconds.
   */
  waitTimeMs?: number;
  /**
   * **Default value**: _<CircularProgress />_
   *
   * React node that will be displayed while either `isLoading` is true,
   * or the elapsed time between `waitTimeMs` and `minDisplayTimeMs` has
   * not passed.
   */
  loadingIndicator?: React.ReactNode;
  /**
   * **Default value**: _<WarningIconError />_
   *
   * React node that will be displayed if an error object
   * is present on `queryRes`.
   */
  errorIndicator?: React.ReactNode;
  /**
   * Query response from a call to `useQuery`
   */
  queryRes: T;
  /**
   * Accepts two child types:
   *   - A callback function that is passed the `data` property from the
   *    `queryRes` once the promise has resolved. If `data` is falsey,
   *    returns the `errorIndicator`
   *   - A regular `React.ReactNode` element
   */
  children:
    | ((resData: Exclude<T['data'], undefined>) => React.ReactNode)
    | React.ReactNode;
}

export const MIN_DISPLAY_TIME_MS = 750;

export const WAIT_TIME_MS = 250;

/**
 * Wraps children with a `minDisplayTimeMs` and `waitTimeMs` props
 * to prevent the loading indicator from flashing on/off screen
 * too quickly.
 *
 * Children will not be rendered until both loading is
 * complete, and the wait timer is no longer active.
 *
 * If loading is completed before the `waitTimeMs` timeout is complete,
 * then no loading indicator will be displayed.
 *
 * If an error occurs, or no data is returned, the `errorIndicator`
 * is displayed.
 */
export const LoadingWithMinDisplay = <T extends QueryResult<any>>({
  minDisplayTimeMs = MIN_DISPLAY_TIME_MS,
  waitTimeMs = WAIT_TIME_MS,
  loadingIndicator = <CircularProgress />,
  errorIndicator = <WarningIconError />,
  queryRes,
  children,
}: LoadingWithMinDisplayProps<T>) => {
  // Used to track updates to the loading state in `useEffect` - this allows us
  // to prevent the loading indicator from being dispalyed if loading completes
  // before `waitTimeMs`.
  const loadingRef = useRef(queryRes.isLoading);

  // Ensures that the loading indicator is displayed for at least `minDisplayTimeMs`
  const [minDisplayTimerActive, setMinDisplayTimerActive] = useState(false);

  // Prevents the loading indicator from displaying for `waitTimeMs`
  const [waitTimerActive, setWaitTimerActive] = useState(true);

  const setDisplayTimeout = () => {
    setMinDisplayTimerActive(true);

    setTimeout(() => {
      setMinDisplayTimerActive(false);
    }, minDisplayTimeMs);
  };

  const setWaitTimeout = () => {
    setTimeout(() => {
      setWaitTimerActive(false);

      if (loadingRef.current) {
        setDisplayTimeout();
      }
    }, waitTimeMs);
  };

  useEffect(() => {
    loadingRef.current = queryRes.isLoading;
    setWaitTimeout();
  }, [queryRes.isLoading]);

  if (waitTimerActive) {
    return null;
  }

  if (minDisplayTimerActive || queryRes.isLoading) {
    return <>{loadingIndicator}</>;
  }

  if (queryRes.error || !queryRes.data) {
    return <>{errorIndicator}</>;
  }

  return typeof children === 'function' ? (
    <>{children(queryRes.data)}</>
  ) : (
    <>{children}</>
  );
};
