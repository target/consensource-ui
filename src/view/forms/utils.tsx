import React from 'react';
import { Grid, Typography, TextFieldProps } from '@material-ui/core';
import { BatchStatusRes } from 'services/api';

export interface TransactionFormProps {
  setBatchStatusLink: (statusLink: BatchStatusRes['link']) => void;
}

export type onChangeEvent = Parameters<
  NonNullable<TextFieldProps['onChange']>
>[0];

/**
 * Helper function that checks if any values in `state` have not been set.
 * Returns true if there is any key in `state` that is falsey.
 */
export function hasEmptyFields<T>(state: T) {
  return Object.keys(state).some((key) => !state[key as keyof T]);
}

export interface FormErrMsgProps {
  msg?: string;
}

export function FormErrMsg({ msg }: FormErrMsgProps) {
  if (!msg) {
    return null;
  }

  return (
    <Grid item>
      <Typography variant="caption" color="error">
        {msg}
      </Typography>
    </Grid>
  );
}
