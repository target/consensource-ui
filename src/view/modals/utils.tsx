import { DialogProps as MUIDialogProps } from '@material-ui/core';
import { FactoryResData } from 'services/api';

export interface DialogProps extends MUIDialogProps {
  handleClose: () => void;
}

export interface OrgDialogProps extends DialogProps {
  existing_org: FactoryResData;
}
