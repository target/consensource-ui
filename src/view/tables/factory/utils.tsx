import { MUIDataTableOptions, MUIDataTableChip } from 'mui-datatables';

export const DEFAULT_ROWS_PER_PAGE = 100;

export const textLabels: MUIDataTableOptions['textLabels'] = {
  body: {
    noMatch: 'No factories found',
  },
};

/**
 * Default props for the filter chips displayed in the
 * `<TableFilterList />`.
 */
export const filterChipProps: MUIDataTableChip = {
  color: 'secondary',
  variant: 'default',
};

/**
 * Utility function to convert a string to an array, or return
 * the original param if an array was passed.
 *
 * This operation is frequent when decoding URL query params.
 */
export const convertStrToArray = (val: string[] | string) => {
  return typeof val === 'string' ? [val] : val;
};
