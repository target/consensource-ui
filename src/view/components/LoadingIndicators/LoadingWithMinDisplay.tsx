import React, { useEffect, useState, useRef, FC } from 'react';
import { LoadingSpinner } from './LoadingSpinner';
import { FullScreenSpinnerSize } from './utils';

export interface LoadingWithMinDisplayProps {
  /**
   * **Default value**: 750
   *
   * Minimum number of milliseconds to display the loading animation for.
   * Note that the timer will only begin once the wait timer has expired.
   *
   */
  minDisplayTimeMs?: number;
  /**
   * **Default value**: 250
   *
   * Number of milliseconds to wait before displaying the loading animation.
   * Used to prevent the loading animation from flashing on screen very briefly
   * if loading completes in less than `waitTimeMs` milliseconds.
   */
  waitTimeMs?: number;
  /**
   * **Default value**: <LoadingSpinner size={FullScreenSpinnerSize} />
   *
   * React node that will be displayed while either `isLoading` is true,
   * or the elapsed time between `waitTimeMs` and `minDisplayTimeMs` has
   * not passed.
   */
  loadingIndicator?: React.ReactNode;
  isLoading: boolean;
}

/**
 * Wraps children with a `minDisplayTimeMs` and `waitTimeMs` props
 * to prevent the loading indicator from flashing on/off screen
 * too quickly.
 *
 * Elements passed as children will not be rendered until both
 * loading is complete, and the timer is no longer active.
 *
 * If loading is completed before the `waitTimeMs` timeout is complete,
 * then no loading indicator will be displayed.
 */
export const LoadingWithMinDisplay: FC<LoadingWithMinDisplayProps> = ({
  minDisplayTimeMs = 750,
  waitTimeMs = 250,
  loadingIndicator = <LoadingSpinner size={FullScreenSpinnerSize} />,
  isLoading,
  children,
}) => {
  const loadingRef = useRef(isLoading);
  const [displayTimerActive, setDisplayTimerActive] = useState(false);
  const [waitTimerActive, setWaitTimerActive] = useState(true);

  const setWaitTimeout = () => {
    setTimeout(() => {
      setWaitTimerActive(false);
      if (loadingRef.current) {
        setDisplayTimerActive(true);
      }
    }, waitTimeMs);
  };

  const setDisplayTimeout = () => {
    setTimeout(() => {
      if (displayTimerActive) {
        setDisplayTimerActive(false);
      }
    }, minDisplayTimeMs + waitTimeMs);
  };

  useEffect(() => {
    loadingRef.current = isLoading;
    setWaitTimeout();
    setDisplayTimeout();
  }, [isLoading]);

  if (waitTimerActive) {
    return null;
  }

  return <>{displayTimerActive ? loadingIndicator : children}</>;
};
