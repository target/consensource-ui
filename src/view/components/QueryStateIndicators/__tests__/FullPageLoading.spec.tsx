import React from 'react';
import { render, act, screen } from 'utils/test-utils';
import { QueryResult } from 'react-query';
import { FullPageLoading } from '../FullPageLoading';
import { WAIT_TIME_MS } from '../LoadingWithMinDisplay';

describe('<FullPageLoading />', () => {
  const baseQueryRes = {
    isLoading: true,
    error: undefined,
    data: [],
  } as QueryResult<any>;

  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    act(() => {
      jest.runAllTimers();
    });
  });

  const advanceTimersInAct = (advanceTimeMs: number) => {
    act(() => {
      jest.advanceTimersByTime(advanceTimeMs);
    });
  };

  it('renders an error indicator with a label', () => {
    const queryRes = { ...baseQueryRes, isLoading: false, error: true };
    const errorLabel = 'error';

    const { container } = render(
      <FullPageLoading queryRes={queryRes} errorLabel={errorLabel}>
        test
      </FullPageLoading>,
    );

    advanceTimersInAct(WAIT_TIME_MS + 1);

    expect(screen.getByTitle(errorLabel)).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('renders a loading indicator with a label', () => {
    const loadingLabel = 'loading';

    const { container } = render(
      <FullPageLoading queryRes={baseQueryRes} loadingLabel={loadingLabel}>
        test
      </FullPageLoading>,
    );

    advanceTimersInAct(WAIT_TIME_MS + 1);

    expect(screen.getByText(loadingLabel)).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
