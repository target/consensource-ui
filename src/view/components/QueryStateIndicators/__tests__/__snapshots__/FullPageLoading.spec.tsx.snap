// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FullPageLoading /> renders a loading indicator with a label 1`] = `
<div>
  <div
    aria-hidden="true"
    class="MuiBackdrop-root makeStyles-backdrop-3"
    style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-direction-xs-column MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
    >
      <div
        class="MuiGrid-root MuiGrid-item"
      >
        <div
          class="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
          role="progressbar"
          style="width: 0px; height: 0px;"
        >
          <svg
            class="MuiCircularProgress-svg"
            viewBox="22 22 44 44"
          >
            <circle
              class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
              cx="44"
              cy="44"
              fill="none"
              r="20.2"
              stroke-width="3.6"
            />
          </svg>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item"
      >
        <h5
          class="MuiTypography-root MuiTypography-h5"
        >
          loading
        </h5>
      </div>
    </div>
  </div>
</div>
`;

exports[`<FullPageLoading /> renders an error indicator with a label 1`] = `
<div>
  <div
    class="makeStyles-centered-2"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 MuiGrid-direction-xs-column MuiGrid-align-items-xs-center"
    >
      <div
        class="MuiGrid-root MuiGrid-item"
      >
        <svg
          class="MuiSvgIcon-root MuiSvgIcon-colorError MuiSvgIcon-fontSizeLarge"
          focusable="false"
          role="img"
          viewBox="0 0 24 24"
        >
          <path
            d="M1 21h22L12 2 1 21zm12-3h-2v-2h2v2zm0-4h-2v-4h2v4z"
          />
          <title>
            error
          </title>
        </svg>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item"
      >
        <h6
          class="MuiTypography-root MuiTypography-h6 MuiTypography-colorError"
        >
          error
        </h6>
      </div>
    </div>
  </div>
</div>
`;
